lib/ddtrace/context.rb

```diff
@@ -89,9 +90,17 @@ def add_span(span)
         # by default has 10000 spans, all of which belong to unfinished parts of a
         # larger trace. This is a catch-all to reduce global memory usage.
         if @max_length > 0 && @trace.length >= @max_length
-          Datadog::Tracer.log.debug("context full, ignoring span #{span.name}")
           # Detach the span from any context, it's being dropped and ignored.
           span.context = nil
+          Datadog::Tracer.log.debug("context full, ignoring span #{span.name}")
+
+          # If overflow has already occurred, don't send this metric.
+          # Prevents metrics spam if buffer repeatedly overflows for the same trace.
+          unless @overflow
+            Diagnostics::Health.metrics.error_context_overflow(1, tags: ["max_length:#{@max_length}"])
```

Comment:
Not sure the tag makes sense here.

I understand wanting to report the max length so you know what value they set, but it might be better off as its own metric, e.g. `dogstatsd.gauge('datadog.tracer.trace.max_spans', @max_length)`

Reply:
Is it because the tag is expensive to add a tag than a metric?

If it were to be decomposed into two metrics, I would think it would be more difficult to accurately correlate the "max spans exceeded" error with the actual limit breached in an environment where there are multiple apps with different max trace lengths. It's conceivable there might be a web app and worker process running on the same host which have different trace lengths and so an event like this would be harder to associate with the appropriate process in order to the max length.

What's the benefit of decomposing this data over having them in one metric?

Reply:
I was looking at how we are reporting another metric, `error_unfinished_spans`, in this same file and that seems to me like a more actionable metric, as it contains the span name affected.

Would something like this work in this case?

```ruby
Health.metrics.error_context_overflow(@trace.length, tags: ["name:#{trace.first.name}"])
```

One concern I have with this approach is that most `trace.first` spans are always going to be `rack.request`, so maybe the value is not really there.

Reply:
That would help address what kind of trace it was in the example I gave, so that's a possibility.

If adding the original tag to the metric is problematic though, I'm not sure how this wouldn't be a problem too (although I don't yet understand why either would be problematic.) In that case, I'd rather opt to remove the tag entirely than create more metrics and make things more complex.